### DISCUSSION TITLE
Is add operation on object reference valid?

### DISCUSSION THREAD

**Discussion by buybackoff** [dotnet/runtime#65793] [Q&A]
ECMA 335 says in one place in parentheses that addition is supported for `O` types.
![image](https://user-images.githubusercontent.com/489852/155381082-6bb3f997-17d6-42d6-9f0d-460f862e17fb.png)

Other places say that's it's not OK. 
![image](https://user-images.githubusercontent.com/489852/155381130-769b3d7c-aa7c-4823-9f4a-a7bad610641f.png)

We have a discussion on the fastest and still correct way to read array elements without bound checks here: https://github.com/disruptor-net/Disruptor-net/issues/65

In the following implementation, is it OK to use the result of `ldind.ref` as a managed pointer and add offset to it? Is the result a GC-tracked managed pointer?

```
  .method public hidebysig static !!0/*T*/&
    UnsafeGetAt<T>(
      !!0/*T*/[] 'array',
      int32 index
    ) cil managed
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.ExtensionAttribute::.ctor()
      = (01 00 00 00 )
    .param type [1] /*T*/
      .custom instance void System.Runtime.CompilerServices.NullableAttribute::.ctor([in] unsigned int8)
        = (01 00 02 00 00 ) // .....
        // unsigned int8(2) // 0x02
    .maxstack 3

    IL_0000: ldarga.s     'array'
    IL_0002: ldind.ref    // Put O on the stack
    IL_0003: ldarg.1      // index
    IL_0004: conv.u      // zero-extend
    IL_0005: sizeof       !!0/*T*/
    IL_000b: mul
    IL_000c: call         int32 ArrayUtil::get_ArrayDataOffset()
    IL_0011: add
    IL_0012: add        // O + offset: the result is tracked & or nint with a GC hole?
    IL_0013: ret           

  } // end of method ArrayUtil::UnsafeGetAt
```

Asking specifically for `#if NETCOREAPP`, on Mono it blows up (and on `net48` it works but it's not that important so could take safe slower path as with Mono and use simple if-def).

**Comment by huoyaoyuan** [comment]
Sounds like what `Unsafe.Add` is doing.

**Reply by buybackoff** [reply]
@huoyaoyuan The first argument of Unsafe.Add is a managed pointer, not an object reference.

**Reply by huoyaoyuan** [reply]
I'm not familiar with precise definition of IL concepts, but AFAIK `Unsafe.Add` is used for span element indexing to bypass bound check.

**Reply by ltrzesniewski** [reply]
`Span<T>` already has something equivalent to a managed pointer (a `ByReference<T>` field, which the runtime treats just like a byref field). This is a `&` type in IL terms, and it allows arithmetic (see the table above).

We want to start with an object reference (`O` type in IL terms) and perform arithmetic on it, which ECMA-335 disallows. The standard also makes it quite clear that the `O` and `&` types are not interchangeable.

The question essentially boils down to: _"how to convert an `O` type to a `&` type in a **supported** way?"_.

**Reply by Joe4evr** [reply]
> The question essentially boils down to: _"how to convert an `O` type to a `&` type in a **supported** way?"_.

Wait for [byref field support to be official](https://github.com/dotnet/runtime/issues/63768)?

**Reply by ltrzesniewski** [reply]
> Wait for [byref field support to be official](https://github.com/dotnet/runtime/issues/63768)?

The question is unrelated to fields. It asks about item types on the IL evaluation stack.

**Comment by SingleAccretion** [comment]
> ECMA 335 says in one place in parentheses that addition is supported for O types.

@buybackoff It is not supported, this seems like a mistake in the standard (though you'll note this is "informative text").

E. g. if I ran the following through RyuJit built with assertions, it'll fail:
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
private static ref int Problem(ClassWithFields a)
{
    IL.Emit.Ldarg(0);
    IL.Emit.Ldc_I4(4);
    IL.Emit.Conv_I();
    IL.Emit.Add();

    return ref IL.ReturnRef<int>();
}
```
```
Assert failure(PID 10764 [0x00002a0c], Thread: 19252 [0x4b34]): Assertion failed 'genActualType(op1->TypeGet()) != TYP_REF && genActualType(op2->TypeGet()) != TYP_REF : Possibly bad IL with CEE_add at offset 0003h (op1=ref op2=long stkDepth=0)' in 'RyuJitReproduction.Program:Problem(ClassWithFields):byref' during 'Importation' (IL size 5)
```
Relying on illegal IL being accepted is undefined behavior and is not supported.

**Reply by ltrzesniewski** [reply]
Thanks! Would you, by any chance, know of a _supported_ way to convert an `O` to an equivalent `&`?

I'm thinking about something like `ldflda <method table "field">`. Currently, we assign an `O` to a `&` local, but something tells me this will also blow up in a RyuJit with assertions...

**Reply by SingleAccretion** [reply]
> Would you, by any chance, know of a supported way to convert an O to an equivalent &?

Nothing that I can think of. The supported way to get `&`s out of `O`s is via `ldflda/ldelema`, of course, that doesn't get you a reference that's exactly the same. Note as well that you cannot use `ldflda/ldelema` with types different from those of the underlying objects.

**Reply by buybackoff** [reply]
> Note as well that you cannot use ldflda/ldelema with types different from those of the underlying objects.

.NET itself does this with `RawArrayData` https://source.dot.net/#System.Private.CoreLib/src/System/Array.CoreCLR.cs,42e9b7616956cf94

@ltrzesniewski  For x-framework (Mono + others) `ldflda` to [`RawData.Data`](https://source.dot.net/#System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs,97aa8714931ebbe9,references) + hardcoded or calculated into SRO offset is the "safe" way.

**Reply by SingleAccretion** [reply]
> .NET itself does this with RawArrayData https://source.dot.net/#System.Private.CoreLib/src/System/Array.CoreCLR.cs,42e9b7616956cf94

Yes, and it has UB issues. See https://github.com/dotnet/runtime/issues/58312#issuecomment-993491291.

**Reply by buybackoff** [reply]
@SingleAccretion  UB in terms of specs or .NET 6 implementation? My original question was if in .NET 6 `O + nint` is `nint` or `&` _in the current implementation_, I could understand that the first picture from specs is a typo.

This is another use of a different type, which is the most relevant to this discussion https://source.dot.net/#System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.CoreCLR.cs,0a97cbc893e44c76,references

**Reply by SingleAccretion** [reply]
>  UB in terms of specs or .NET 6 implementation?

Both.

**Reply by SingleAccretion** [reply]
> My original question was if in .NET 6 `O + nint` is `nint` or `&` in the current implementation

I understand what you're getting at -- and I can only tell you that internally the compiler treats  `O + nint` as `&`. However, as the assertion showed above tells - it is unknown and untested how will the compiler behave if such constructs appear in user code (IL). It may behave sensibly and reliably, it may crash, it may silently produce bad (in subtle ways) code.

**Reply by buybackoff** [reply]
I have 700 unit tests all depending on that, everything works well. But on a quiet machine without much GC (but tests are running in parallel). Maybe I should leave overnight "run until failure", but with your comment this will not give anything. I really hoped that in the current implementation there is no difference between O and &.

For GC it looks safe from this comment: https://github.com/dotnet/coreclr/pull/20386#issuecomment-434964787
